module/src/hooks/clientside-validation/formValidationHook.tsx

```diff
@@ -0,0 +1,132 @@
+/**
+ * Hook to enable real time validation of an armstrong form. The form validation is defined in `inputValidators`, and
+ * the form state in `state`. This returns an object containing the following:
+ *  - `setTouched`: a method provided to tell when a form input control has been interacted with by the user. This must
+ * be used on each input with validation. As an example:
+ *
+ * ```
+ * <TextInput {...bind.text('city')} label="City" onBlur={setTouched} />
+ * ```
+ *
+ *  - `formIsValid`: a boolean flag to denote if the form is valid or invalid. This doesn't rely on user interaction, so
+ * is independant of error messages displayed on the form.
+ *  - `validationErrors`: an array populated with the errors to display on the form. Should be passed to the `DataForm` component:
+ *
+ * ```
+ * <DataForm validationResults={validationErrors} validationMode="both">
+ * ```
+ *
+ * @param inputValidators form validation
+ * @param dataBinder form state
+ */
+export function useValidatedForm<TData>(inputValidators: InputValidator[], state: TData, formProp: (...keyChain: KeyChain) => BindingTools<TData>) {
+  // Use this to calculate if this is the first render. If it is, then don't run the validation useEffect as it'll show all of the validation errors prematurely.
+  const firstRenderCheck = React.useRef(0);
+
+  // Run validation on individual inputs whenever they change value.
+  for (const inputValidator of inputValidators) {
+    React.useEffect(() => {
+      if (firstRenderCheck.current < inputValidators.length) {
+        firstRenderCheck.current++;
+        return;
+      }
+      runValidationOnInput(inputValidator, state, formProp);
+    }, [valueByKeyChain(state, inputValidator.key)]);
+  }
+
+  // Display all validation error messages on the form
+  function showAllValidation() {
+    for (const inputValidator of inputValidators) {
+      runValidationOnInput(inputValidator, state, formProp);
+    }
+  }
+
+  const formIsValid = calculateFormValidity(state, inputValidators);
+
+  // Everything to do with input state being touched below:
+  const [inputStates, setInputStates] = React.useState<InputState[]>(
+    inputValidators.map(input => ({
+      key: input.key,
+      touched: false,
+    }))
+  );
+
+  const setTouched = (inputElement: React.FocusEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>) => {
+    const inputStatesCopy = [...inputStates];
+    const touchedInput = inputStatesCopy.find(
+      // TODO doesn't work as the name property isn't specified on inputs anymore.
+      input => convertKeyChainToDotNotation(input.key) === inputElement.target.name
+    );
+    if (touchedInput) {
+      touchedInput.touched = true;
+      setInputStates(inputStatesCopy);
+    }
+  };
+
+
+  // Run validation on individual inputs whenever their state changes to 'touched'.
+  // This covers the scenario where a user clicks on an input, doesn't change anything, and clicks off.
+  for (const inputState of inputStates) {
```

Comment:
same as above, it's definitely dangerous to put hooks inside loops

Reply:
I'll look into doing it without the loop. Worth noting, as above, that it is not possible for the length of inputStates to change